data_packet.cpp

```diff
@@ -118,4 +120,65 @@ size_t StateBuffer::get_id_size() {
         return 0;
     }
 }
+
+bool StateBuffer::operator==(const StateBuffer& other) {
+    if (reals_ && !other.reals_ || !reals_ && other.reals_ ||
+        reals_ && other.reals_ && *reals_ != *other.reals_) {
+        return false;
+    }
+    if (pixels_ && !other.pixels_ || !pixels_ && other.pixels_ ||
+        pixels_ && other.pixels_ && *pixels_ != *other.pixels_) {
+        return false;
+    }
+    if (id_ && !other.id_ || !id_ && other.id_ ||
+        id_ && other.id_ && *id_ != *other.id_) {
+        return false;
+    }
+    if (str_ && !other.str_ || !str_ && other.str_ ||
+        str_ && other.str_ && *str_ != *other.str_) {
+        return false;
+    }
+
+    return true;
 }
+
+void StateBuffer::encode(util::BinaryBuffer& buf) {
+    uint8_t flags = (reals_ ? 1 : 0) | (pixels_ ? 2 : 0) |
+                    (id_ ? 4 : 0) | (str_ ? 8 : 0);
```

Comment:
It might be better to use a enum type for the flags (1, 2, 4 ,8) both here and in line 166-178?